core/src/main/java/org/apache/accumulo/core/spi/crypto/AESCryptoService.java

```diff
@@ -132,11 +134,11 @@ public FileEncrypter getFileEncrypter(CryptoEnvironment environment) {
   @Override
   public FileDecrypter getFileDecrypter(CryptoEnvironment environment) {
     CryptoModule cm;
-    byte[] decryptionParams = environment.getDecryptionParams();
-    if (decryptionParams == null || checkNoCrypto(decryptionParams))
+    var decryptionParams = environment.getDecryptionParams();
+    if (decryptionParams.isEmpty() || checkNoCrypto(decryptionParams.get()))
       return new NoFileDecrypter();
 
-    ParsedCryptoParameters parsed = parseCryptoParameters(decryptionParams);
+    ParsedCryptoParameters parsed = parseCryptoParameters(decryptionParams.get());
     Key kek = loadDecryptionKek(parsed);
     Key fek = unwrapKey(parsed.getEncFek(), kek);
```

Comment:
What does `Fek` mean in this context?

Reply:
File encrypting key? I am not sure its been a while.

Reply:
Yeah, some of the abbreviations are foreign to me. I think you are right that `fek` is [File Encryption Key](https://csrc.nist.gov/glossary/term/fek). However, it's not used for the entire file, right? It's used for the block? I have seen it referenced as a [Secret Key](https://csrc.nist.gov/glossary/term/secret_key_cryptographic_algorithm) in the [JCA](https://docs.oracle.com/javase/8/docs/technotes/guides/security/crypto/CryptoSpec.html#SecretKeyFactory).

Reply:
It is used for both, the WALs or block encryption for R-Files, depending on which Scope is provided in the `CryptoEnvironment`.
https://github.com/apache/accumulo/blob/41310c57dbf840ab0ed5faea70f6c3b8311a9697/core/src/main/java/org/apache/accumulo/core/spi/crypto/CryptoService.java#L44-L48

Reply:
To clarify... We use "AES/CBC/NoPadding" transformation for WALs and "AES/GCM/NoPadding" transformation for RFiles. So different modes per type of file.

Reply:
I take that back, it looks like it might be used for the entire file and a new IV is generated for each block. I do have a concern though that we are leaving unencrypted secret keys (fek) in the heap after using them. Reference [here](https://github.com/apache/accumulo/blob/d454afd39c91ee0a1ab267e0caafc52254d209b0/core/src/main/java/org/apache/accumulo/core/spi/crypto/AESCryptoService.java#L316), [here](https://github.com/apache/accumulo/blob/d454afd39c91ee0a1ab267e0caafc52254d209b0/core/src/main/java/org/apache/accumulo/core/spi/crypto/AESCryptoService.java#L449) and [here](https://github.com/apache/accumulo/blob/d454afd39c91ee0a1ab267e0caafc52254d209b0/core/src/main/java/org/apache/accumulo/core/spi/crypto/AESCryptoService.java#L141). I believe, looking at the class javadoc for [SecretKey](https://github.com/openjdk/jdk/blob/jdk-11%2B27/src/java.base/share/classes/javax/crypto/SecretKey.java), we should be calling `destroy()` when we are done with it. I'm not sure we are doing that.

Reply:
Are you talking about these changes or the current code?

Reply:
FEK = file encrypting key, KEK = key encrypting key. A Google search resulted in: https://en.wikipedia.org/wiki/Glossary_of_cryptographic_keys
I've heard "FEK" more commonly than "DEK", but "DEK" seems to be a more generic term for basically the same thing... as far as I can tell from that glossary anyway. "FEK" is used in https://en.wikipedia.org/wiki/Encrypting_File_System and https://www.kernel.org/doc/html/latest/security/keys/ecryptfs.html, so I think it's more common in terms of file system or file-based encryption.

In any case, the idea is that one key that's relatively persistent somewhere (KEK) encrypts a random key (FEK) that's used to actually encrypt the file's blocks. I think Dave is correct the FEK is used for the whole file, with each block using a new IV. The KEK should be per-table.

I agree that we should call destroy on them after we're done with them. I'm not sure it's worth zeroing out the byte arrays manually, though, because that's making a lot of assumptions about how Java handles memory internally, and I think it's not likely to work the way we expect. Calling destroy on them should be easy, and we would be deferring to the implementation to do it correctly.

Reply:
I can try to integrate the call to destroy into this PR, pulling some of Dave's changes from #2616 . Should I wait until you take a look at my latest changes?